src/utils/gitlab_updater.rs

```diff
@@ -0,0 +1,396 @@
+use crate::error::Result;
+use crate::{db::Pool, Config};
+use chrono::{DateTime, Utc};
+use log::{info, trace, warn};
+use once_cell::sync::Lazy;
+use postgres::Client;
+use regex::Regex;
+use reqwest::{
+    blocking::Client as HttpClient,
+    header::{HeaderMap, HeaderValue, ACCEPT, AUTHORIZATION, USER_AGENT},
+};
+use serde::Deserialize;
+use std::collections::{HashMap, HashSet};
+use std::sync::Arc;
+
+use crate::utils::{RepositoryName, Updater, APP_USER_AGENT};
+
+const GRAPHQL_UPDATE: &str = "query($ids: [ID!]!) {
+    projects(ids: $ids) {
+        nodes {
+            id
+            nameWithNamespace
+            lastActivityAt
+            description
+            starCount
+            forksCount
+            openIssuesCount
+        }
+    }
+}";
+
+const GRAPHQL_SINGLE: &str = "query($fullPath: ID!) {
+    project(fullPath: $fullPath) {
+        id
+        nameWithNamespace
+        lastActivityAt
+        description
+        starCount
+        forksCount
+        openIssuesCount
+    }
+}";
+
+const ALLOWED_HOSTS: &[&str] = &["gitlab.com", "gitlab.freedesktop.org"];
+
+/// How many repositories to update in a single chunk.
+const UPDATE_CHUNK_SIZE: usize = 5;
+
+fn extract_host(url: &str) -> Option<&str> {
+    url.split("//").nth(1).and_then(|u| u.split('/').next())
+}
+
+pub struct GitlabUpdater {
+    client: HttpClient,
+    pool: Pool,
+}
+
+impl Updater for GitlabUpdater {
+    /// Returns `Err` if the access token has invalid syntax (but *not* if it isn't authorized).
+    /// Returns `Ok(None)` if there is no access token.
+    fn new(config: Arc<Config>, pool: Pool) -> Result<Option<Self>> {
+        let mut headers = HeaderMap::new();
+        headers.insert(USER_AGENT, HeaderValue::from_static(APP_USER_AGENT));
+        headers.insert(ACCEPT, HeaderValue::from_static("application/json"));
+
+        if let Some(token) = &config.gitlab_accesstoken {
+            headers.insert(
+                AUTHORIZATION,
+                HeaderValue::from_str(&format!("Bearer {}", token))?,
+            );
+        } else {
+            warn!("will try to retrieve Gitlab stats without token since none was provided");
+        }
+
+        let client = HttpClient::builder().default_headers(headers).build()?;
+
+        Ok(Some(GitlabUpdater { client, pool }))
+    }
+
+    fn backfill_repositories(&self) -> Result<()> {
+        info!("started backfilling Gitlab repository stats");
+
+        let mut conn = self.pool.get()?;
+        let needs_backfilling = conn.query(
+            "SELECT releases.id, crates.name, releases.version, releases.repository_url
+             FROM releases
+             INNER JOIN crates ON (crates.id = releases.crate_id)
+             WHERE repository IS NULL AND repository_url LIKE '%gitlab.%/%';",
```

Comment:
This url check is wrong for any gitlab based forge without gitlab in the hostname. Instead of having all gitlab forges checked in a single instance we could do something like 

```rust
for host in GITLAB_HOSTS {
  GitlabUpdater::new(config, pool, host).backfill_repositories();
}
```

i.e. treat each hostname as a fully separate instance.

Reply:
Oh wow. Great idea!